File: leetcode/python/0278_first_bad_version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 1: return 1

        # 1 2 3 4 5
        # F F F T T
        l, r, middle = 1, n, 0

        while l <= r:
            middle = l + (r - l) // 2
            res = isBadVersion(middle)

            if res == True:
                r = middle - 1
            elif res == False:
                l = middle + 1

        return l

        

File: leetcode/java/0278_first_bad_version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        if (n == 1) return n;
        int l = 1, r = n, middle = 0;

        while (l <= r) {
            middle = l + (r - l) / 2;
            boolean bad = isBadVersion(middle);

            // false false false true true
            // 1       2      3    4   5

            // if bad == false: l = middle + 1
            // if bad == true: r = middle - 1
            if (bad == false) l = middle + 1;
            if (bad == true) r = middle - 1;
        }

        return l;
    }
}